src/python/WMCore/Storage/Backends/GFAL2Impl.py

```diff
@@ -113,27 +114,78 @@ def createStageOutCommand(self, sourcePFN, targetPFN, options=None, checksums=No
                 copyCommandDict['checksum'] = "-K adler32"
 
         copyCommandDict['options'] = ' '.join(unknown)
-
         copyCommandDict['source'] = self.createFinalPFN(sourcePFN)
         copyCommandDict['destination'] = self.createFinalPFN(targetPFN)
 
         copyCommand = self.copyCommand % copyCommandDict
         result += copyCommand
 
+        logging.debug("Actual command which failed: %s", copyCommand)
+
         if _CheckExitCodeOption:
             result += """
             EXIT_STATUS=$?
             echo "gfal-copy exit status: $EXIT_STATUS"
             if [[ $EXIT_STATUS != 0 ]]; then
-               echo "ERROR: gfal-copy exited with $EXIT_STATUS"
-               echo "Cleaning up failed file:"
-               %s
+                echo "ERROR: gfal-copy exited with $EXIT_STATUS"
+                echo "Source PFN: {source}"
+                echo "Target PFN: {destination}"
+                echo "Cleaning up failed file: {remove_command}"
             fi
             exit $EXIT_STATUS
-            """ % self.createRemoveFileCommand(targetPFN)
+            """.format(
+                source=copyCommandDict['source'], 
+                destination=copyCommandDict['destination'], 
+                remove_command=self.createRemoveFileCommand(targetPFN)
+            )
+        
+        return result
+
+    def createDebuggingCommand(self, sourcePFN, targetPFN, options=None, checksums=None):
+        """
+        Report environment information about a failed gfal-copy command
+        """
+        result = "#!/bin/bash\n"
+
+        copyCommandDict = {'checksum': '', 'options': '', 'source': '', 'destination': ''}
```

Comment:
AFAICT, the actual construction of the stage out command (without any of the exit status check and file removal) is exactly the same between this method and `createStageOutCommand`. If that is correct, then I think we should separate that logic in a different method and call that method inside `createStageOutCommand()` or `createDebuggingCommand()`. This way we guarantee that the code/command will never diverge by accident.

Another option, but it is a larger step, would be to save the actual stage out command (without exit status check/file removal overload) in a instance variable and just access that inside `createDebuggingCommand()`. This gives us even less room for potential differences in the command construction for real execution and debugging execution (plus, it avoids many internal calls).